### PR TITLE
Stricter types for Chrome APIs

### PR DESCRIPTION
**Pull request by bershanskiy**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: will link individual URLs for every change as a comment
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:499]: See comment above. Also, docs:
 - https://developer.chrome.com/docs/extensions/reference/bookmarks/#type-BookmarkTreeNodeUnmodifiable
 - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/BookmarkTreeNodeUnmodifiable

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:6905]: This API was never shipped in release browsers. [(It was a Chrome-only experiment which was later retracted.)](https://stackoverflow.com/questions/20306548/how-to-use-system-indicator-in-the-manisfet-of-a-chrome-packaged-app)

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:10038]: See comment above and docs: https://developer.chrome.com/docs/extensions/reference/wallpaper/#type-WallpaperLayout

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:10031]: https://developer.chrome.com/docs/extensions/reference/wallpaper/#method-setWallpaper

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:10050]: https://developer.chrome.com/docs/extensions/reference/wallpaper/#method-setWallpaper

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:798]: Docs: https://developer.chrome.com/docs/extensions/reference/browserAction/#method-setIcon

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:1592]: https://developer.chrome.com/docs/extensions/reference/contextMenus/#method-create

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:2090]: https://developer.chrome.com/docs/extensions/reference/declarativeWebRequest/#type-HeaderFilter

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:3198]: https://developer.chrome.com/docs/extensions/reference/events/#type-UrlFilter

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:4300]: Docs:
 - https://developer.chrome.com/docs/extensions/reference/i18n/#method-getMessage
 - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/i18n/getMessage

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:5785]: https://developer.chrome.com/docs/extensions/reference/pageCapture/#method-saveAsMHTML

File: types/chrome/index.d.ts
Comment (bershanskiy) [inline types/chrome/index.d.ts:6003]: It's an object representing printer capabilities in CDD format.

**Comment by nicolas377** [comment]
Just make an empty commit here, this isn't your fault.

File: types/chrome/index.d.ts
Comment (nicolas377) [reply types/chrome/index.d.ts:1592]: Since the `exactOptionalPropertyTypes` typescript flag is set, undefined should be explicitly declared.
```suggestion
        parentId?: number | string | undefined;
```

File: types/chrome/index.d.ts
Comment (nicolas377) [reply types/chrome/index.d.ts:2090]: Since the `exactOptionalPropertyTypes` typescript flag is set, undefined should be explicitly declared.
```suggestion
        valueContains?: string | string[] | undefined;
```

File: types/chrome/index.d.ts
Comment (nicolas377) [inline types/chrome/index.d.ts:2094]: Since the `exactOptionalPropertyTypes` typescript flag is set, undefined should be explicitly declared.
```suggestion
        nameContains?: string | string[] | undefined;
```

File: types/chrome/index.d.ts
Comment (nicolas377) [reply types/chrome/index.d.ts:3198]: Since the `exactOptionalPropertyTypes` typescript flag is set, undefined should be explicitly declared.
```suggestion
        ports?: (number | number[])[] | undefined;
```

File: types/chrome/index.d.ts
Comment (nicolas377) [inline types/chrome/index.d.ts:5714]: Since the `exactOptionalPropertyTypes` typescript flag is set, undefined should be explicitly declared.
```suggestion
        path?: string | { [index: number]: string } | undefined;
```

File: types/chrome/index.d.ts
Comment (nicolas377) [reply types/chrome/index.d.ts:6003]: `object` and its sister type, `Object`, are very ambiguous types, as pointed out [here](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/ban-types.md#default-options). I'm not fully sure what you're going for here, but here's the error message that typescript-eslint will give for some guidance:
```
The `Object` type actually means "any non-nullish value", so it is marginally better than `unknown`.
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
```

File: types/chrome/index.d.ts
Comment (nicolas377) [inline types/chrome/index.d.ts:9855]: This should also be extracted to a type, as it is also used at line 9872.

File: types/chrome/index.d.ts
Comment (nicolas377) [reply types/chrome/index.d.ts:10031]: Since the `exactOptionalPropertyTypes` typescript flag is set, undefined should be explicitly declared.
```suggestion
        data?: ArrayBuffer | undefined;
```

File: types/chrome/index.d.ts
Comment (nicolas377) [reply types/chrome/index.d.ts:10038]: Since the `exactOptionalPropertyTypes` typescript flag is set, undefined should be explicitly declared.
```suggestion
        layout: "STRETCH" | "CENTER" | "CENTER_CROPPED" | undefined;
```

**Review by nicolas377 — CHANGES_REQUESTED** [review]
There's one big thing I wasn't able to cram into a specific comment. In the contentSettings API, `SetDetails` has changed to `ContentSetting`, as well as in all the specific names (like `ImagesSetDetails` => `ImagesContentSetting`. If that's beyond the scope of this PR, I'd be happy to approve this after the rest of the changes I noted are implemented, and take care of the transition from `SetDetails` to `ContentSetting` myself.

**Comment by bershanskiy** [comment]
Thanks for the review, I pushed more commits: the first one addresses your feedback, the other two just polish up some other rough edges.

**Review by nicolas377 — APPROVED** [review]
LGTM! Once this gets merged, I'll open a PR to migrate away from the `ContentSetting`.

**Comment by bershanskiy** [comment]
Should I do anything else to advance this PR? Asking just in case I missed something.

**Comment by nicolas377** [comment]
Since there are no tests for the PR, we need a maintainer to approve this PR. @elibarzilay seems to have given a blessing, could you take a look at this PR again?

**Pull request merged**
